### DISCUSSION TITLE
Get payment using Connect and Test Env

### DISCUSSION THREAD

**Discussion by mschewe** [VictorAvelar/mollie-api-go#219] [Q&A] [answered]
Hello, 

thanks for all the effort to create this library :+1: :pray:!

I am currently implementing Mollie with it. I think I came around one edge case that seems not to work properly. 

1. Create a payment using oAuth / in `testMode`
2. Receive the webhook and try to load the transaction using oAuth

I would expect that it would try to retrieve the transaction with the `testMode` enabled. 
Somehow it seems as if the client does not use the `testMode`.

I assume this behavior from this error message:

```
404 Not Found: Payment tr_XXXXXX exists, but the wrong mode is used. Try switching live / test API keys."
```

This is how I create the client for every request:

```go
// NewMollieService creates a new mollie client instance with the given token
func (m mollieService) newClient(token oauth2.Token) (*mollie.Client, error) {

	tokenSource := m.authConfig.TokenSource(context.Background(), &token)
	baseClient := oauth2.NewClient(context.Background(), tokenSource)

	logrus.WithField("testMode", m.testMode).Info("testMode") // testMode=true
	config := mollie.NewConfig(m.testMode, false, "NOT_NEEDED_HERE?")

	mollieClient, err := mollie.NewClient(baseClient, config)
	if err != nil {
		logrus.WithError(err).Error("could not create mollie mollieClient")
		return nil, err
	}
	return mollieClient, nil
}
```
I am not sure, if this might be a bug or if I am using the client wrong.
Is there an easy way to ensure the client is running in testMode?

Thank you very much for your help!

**Comment by VictorAvelar** [comment, marked as answer]
Hello @mschewe, I appreciate you bringing this issue to our attention. I have a few theories and suggestions to: 1. Verify that it is operating in test-mode (which, as of right now, I believe it is not); and 2. Resolve the connect client problem.

### Verify that it is operating in test-mode

```go
// client initialization

tokenSource := m.authConfig.TokenSource(context.Background(), &token)
baseClient := oauth2.NewClient(context.Background(), tokenSource)

logrus.WithField("testMode", m.testMode).Info("testMode") // testMode=true
config := mollie.NewConfig(m.testMode, false, "NOT_NEEDED_HERE?")

mollieClient, err := mollie.NewClient(baseClient, config)
if err != nil {
  logrus.WithError(err).Error("could not create mollie mollieClient")
  return nil, err
}

// payment creation
res, payment, err := mollieClient.Payments.Create(context.Background(), &mollie.Payment{/*required values */}, nil)
if err != nil {
    logrus.WithError(err).Error("payment creation failed")
}

// Check correct values are set

logrus.WithField("query-values", res.Request.URL.Query().Get("testmode")).Info("testmode query param") // should be true and is set by the global client config
logrus.WithField("payment-mode", payment.TestMode).Info("testmode inside payment") // should also be true and is set by the `PaymentsService.Create` method.
```

### Theories on why it is failing

1. The `profileId` or `testmode` query parameters are absent when creating payments. [See](https://docs.mollie.com/reference/v2/payments-api/create-payment#access-token-parameters)
2. Due to a bug, the TestMode flag inside the payment struct isn't set; this value isn't required when using API tokens, and the present implementation only handles access tokens. [See](https://github.com/VictorAvelar/mollie-api-go/blob/master/mollie/payments.go#L157).

This calls for either writing a method to handle the OAuth use case (similar to [`client.HasAccessTokens`](https://github.com/VictorAvelar/mollie-api-go/blob/1ef5d0365f274b02881f5e55c349d43facf17255/mollie/mollie.go#L151)), alternatively, one may modify the settings to explicitly manage the OAuth flow before using it to modify [`PaymentsService.Create`](https://github.com/VictorAvelar/mollie-api-go/blob/master/mollie/payments.go#L157)).
